### PR TITLE
PT-316: Add breadcrumbs into products query response schema

### PR DESCRIPTION
**Pull request by AliveMen**
Add two breadcrumbs types for category and product
Add two extensions to calculate breadcrumbs
Add the field 'breadcrumbs' to category and product graph types

**Jira task**: https://virtocommerce.atlassian.net/browse/PT-316

### CODE REVIEW

**Review by eMazeika — APPROVED** [review]

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-7276 has been created

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/ExpProduct.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Security.Policy;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/ExpProduct.cs:4]: Why we need this?

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/CategoryExtension.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using VirtoCommerce.StoreModule.Core.Model;
+using VirtoCommerce.XDigitalCatalog.Breadcrumbs;
+
+namespace VirtoCommerce.XDigitalCatalog.Extensions
+{
+    public static class CategoryExtension
+    {
+        public static IEnumerable<Breadcrumb> GetBreadcrumbs(this Category category, Store store, string cultureName)
+        {
+            if (category.Outlines == null)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/CategoryExtension.cs:15]: Check `category` for null, please.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using VirtoCommerce.StoreModule.Core.Model;
+using VirtoCommerce.XDigitalCatalog.Breadcrumbs;
+
+namespace VirtoCommerce.XDigitalCatalog.Extensions
+{
+    public static class ProductExtension
+    {
+        public static IEnumerable<Breadcrumb> GetBreadcrumbs(this CatalogProduct product, Store store, string cultureName)
+        {
+            if (product.Category != null)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs:13]: Invert if, please.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/CategoryMappingProfile.cs [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
         {
-            CreateMap<SearchDocument, ExpCategory>().ConvertUsing(src => new GenericModelBinder<ExpCategory>().BindModel(src) as ExpCategory);
+            CreateMap<SearchDocument, ExpCategory>().ConvertUsing((src, dest, context) =>
+            {
+                var expCategory = new GenericModelBinder<ExpCategory>().BindModel(src) as ExpCategory;
+
+                if (expCategory != null)
+                {
+                    expCategory.Store = context.Options.Items["store"] as Store;
+                    expCategory.CultureName = context.Options.Items["language"].ToString();
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/CategoryMappingProfile.cs:19]: Change `language` to `cultureName`, please.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            var query = parameter.Query;
+            if (query == null)
+            {
+                throw new OperationCanceledException("Query must be set");
+            }
+
+            if (query.IncludeFields.ContainsAny("breadcrumbs"))
+            {
+                var categoryIds = parameter.Results.Select(x => x.IndexedProduct.CategoryId).ToArray();
+                var categories = await _categoryService.GetByIdsAsync(categoryIds, string.Join(",", CategoryResponseGroup.WithParents.ToString(), CategoryResponseGroup.WithOutlines.ToString()));
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs:39]: Move this code
``` csharp
string.Join(",", CategoryResponseGroup.WithParents.ToString(), CategoryResponseGroup.WithOutlines.ToString())
```
to private method
``` csharp
private string JoinResponseGroups(param CategoryResponseGroup[] responseGroups);
```

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+            var query = parameter.Query;
+            if (query == null)
+            {
+                throw new OperationCanceledException("Query must be set");
+            }
+
+            if (query.IncludeFields.ContainsAny("breadcrumbs"))
+            {
+                var categoryIds = parameter.Results.Select(x => x.IndexedProduct.CategoryId).ToArray();
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs:38]: Should we distinct here?

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                var categories = await _categoryService.GetByIdsAsync(categoryIds, string.Join(",", CategoryResponseGroup.WithParents.ToString(), CategoryResponseGroup.WithOutlines.ToString()));
+
+                foreach (var product in parameter.Results)
+                {
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs:42]:
``` csharp
if (product.IndexedProduct is null) continue;
```

Comment (Vectorfield4) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs:42]: Add this if this is not redundant.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Schemas/BreadcrumbsType.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    public class BreadcrumbType : ObjectGraphType<Breadcrumb>
+    {
+        public BreadcrumbType()
+        {
+            Name = "Breadcrumb";
+
+            Field(x => x.Title).Description("Name of current breadcrumb");
+            Field(x => x.TypeName, nullable: true).Description("Catalog, category or product");
+            Field(x => x.SeoPath, nullable: true).Description("Full path from catalog");
+            Field(x => x.Url, nullable: true).Description("Url of the breadcrumb");
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Schemas/BreadcrumbsType.cs:15]: Breadcrumb without a link? Is this possible from a business perspective?

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]
Please, fix code smells and check comments.

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-7277 has been created

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/CategoryExtension.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/CategoryExtension.cs:20]: Done

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs (hunk 3 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs:13]: I can't. ProductBreadcrumb should be calculated in all cases.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/CategoryMappingProfile.cs (hunk 4 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/CategoryMappingProfile.cs]: done .

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs (hunk 5 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs]: done

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs (hunk 7 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs:40]: I suggest using the "where" condition in foreach

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Schemas/BreadcrumbsType.cs (hunk 8 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Schemas/BreadcrumbsType.cs:15]: URL property removed. Discussed with ET.

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs (hunk 6 above)
Comment (AliveMen) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsBreadcrumbsMiddleware.cs]: added

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+            {
+                foreach (var breadCrumb in product.Category.GetBreadcrumbs(store, cultureName).Distinct())
+                {
+                    yield return breadCrumb;
+                }
+            }
+            yield return new ProductBreadcrumb(product)
+            {
+
```
Comment (ksavosteev) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Extensions/ProductExtension.cs:22]: remove empty line

**Review by Vectorfield4 — APPROVED** [review]

**Comment by AliveMen** [comment]
https://github.com/VirtoCommerce/vc-module-experience-api/pull/126

**Pull request closed without merging**